documentation/design_docs/L1/system_controller.md

```diff
@@ -0,0 +1,154 @@
+# System Controller Interface
+
+- [System Controller Interface](#system-controller-interface)
+- [Location](#location)
+- [Type](#type)
+- [Background](#background)
+- [Overview](#overview)
+- [Detailed Design](#detailed-design)
+  - [Interface](#interface)
+  - [Configuring the Default System Controller for the Platform](#configuring-the-default-system-controller-for-the-platform)
+    - [static void SetPlatformController(SystemController * system_controller)](#static-void-setplatformcontrollersystemcontroller--systemcontroller)
+    - [static SystemController & GetPlatformController()](#static-systemcontroller--getplatformcontroller)
+  - [Clock Configuration](#clock-configuration)
+    - [void * GetClockConfiguration()](#void--getclockconfiguration)
+    - [ClockConfiguration & GetClockConfiguration()](#clockconfiguration--getclockconfiguration)
+  - [Initialization & Clock Configuration](#initialization--clock-configuration)
+    - [Returns&lt;void> Initialize()](#returnsvoid-initialize)
+  - [Getting the Clock Rate of a Resource](#getting-the-clock-rate-of-a-resource)
+    - [Returns&lt;units::frequency::hertz_t> GetClockRate(PeripheralID peripheral) const](#returnsunitsfrequencyhertzt-getclockrateperipheralid-peripheral-const)
+  - [Peripheral Power Control](#peripheral-power-control)
+    - [Returns&lt;bool> IsPeripheralPoweredUp(PeripheralID peripheral) const](#returnsbool-isperipheralpoweredupperipheralid-peripheral-const)
+    - [Returns&lt;void> PowerUpPeripheral(PeripheralID peripheral) const](#returnsvoid-powerupperipheralperipheralid-peripheral-const)
+    - [Returns&lt;void> PowerDownPeripheral(PeripheralID peripheral) const](#returnsvoid-powerdownperipheralperipheralid-peripheral-const)
+- [Future Advancements](#future-advancements)
+- [Testing Plan](#testing-plan)
+  - [Unit Testing Scheme](#unit-testing-scheme)
+
+# Location
+`L1 Peripheral`
+
+# Type
+`Interface`
+
+# Background
+A SystemController manages a platform's:
+
+  - Clocks and their speeds
+  - PLLs feeding into the system's clocks
+  - Manages power systems of peripherals and system blocks
+
+The SystemController also gives information about the system such as what
+speed a clock is running at so that their peripheral drivers can calculate
+things such as the clock rates for serial communicate or the clock cycles before
+an event or interrupt is planned to occur.
+
+# Overview
+Every SystemController should take a custom `ClockConfiguration` data
+structure reference as an input in their constructor, unless a system does
+not incorporate any form of clock tree.
+
+> See [Clock Control Guide](https://sjsu-dev2.readthedocs.io/en/latest/guides/clock_systems)
+> for more information regarding the use of `ClockConfiguration`.
+
+The SystemController uses the `PeripheralID` helper class to define various
```

Comment:
Would it be better to name **PeripheralID** as **ResourceID**?

Reply:
Yeah, that sounds like a good generic name for that. Good idea!